tket/src/Circuit/DAGProperties.cpp

```diff
@@ -107,31 +122,121 @@ bool is_valid(const DAG &G) {
+
+    CHECK(c_in_ports.size() == c_in_deg);
+    CHECK(c_out_ports.size() == c_out_deg);
+
+    CHECK(b_in_ports.size() == b_in_deg);
+
+    if (c_in_empty && c_out_empty && q_in_empty && !q_out_empty) {
+      // quantum in vertex
+      CHECK(b_in_empty);
+      CHECK(b_out_empty);
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(q_in_deg == 0 && q_out_deg == 1);
+    } else if (c_in_empty && c_out_empty && !q_in_empty && q_out_empty) {
+      // quantum out vertex
+      CHECK(b_in_empty);
+      CHECK(b_out_empty);
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(q_in_deg == 1 && q_out_deg == 0);
+    } else if (c_in_empty && c_out_empty && !q_in_empty && !q_out_empty) {
+      // quantum vertex
+      CHECK(b_out_empty);
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(q_in_ports == q_out_ports);  // bijection between in and out ports
+    } else if (w_in_empty && !w_out_empty) {
+      // wasm out vertex
+      CHECK(b_in_empty);
+      CHECK(b_out_empty);
+      CHECK(c_in_empty);
+      CHECK(c_out_empty);
+      CHECK(q_in_empty);
+      CHECK(q_out_empty);
+      CHECK(w_in_deg == 0 && w_out_deg == 1);
+    } else if (!w_in_empty && w_out_empty) {
+      // wasm in vertex
+      CHECK(b_in_empty);
+      CHECK(b_out_empty);
+      CHECK(c_in_empty);
+      CHECK(c_out_empty);
+      CHECK(q_in_empty);
+      CHECK(q_out_empty);
+      CHECK(w_in_deg == 1 && w_out_deg == 0);
+    } else if (!w_in_empty && !w_out_empty) {
+      // wasm vertex
+      CHECK(q_in_empty);
+      CHECK(q_out_empty);
+      CHECK(c_in_ports == c_out_ports);  // bijection between in and out ports
+      CHECK(w_in_ports == w_out_ports);  // bijection between in and out ports
+    } else if (c_in_empty && !c_out_empty) {
+      // claccical in vertex, can have bool out edges
+      CHECK(b_in_empty);
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(q_in_empty);
+      CHECK(q_out_empty);
+      CHECK(c_in_deg == 0 && c_out_deg == 1);
+    } else if (!c_in_empty && c_out_empty) {
+      // claccical out vertex, can have bool in edges
+      CHECK(b_out_empty);
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(q_in_empty);
+      CHECK(q_out_empty);
+      CHECK(c_in_deg == 1 && c_out_deg == 0);
+    } else if (q_in_empty && q_out_empty && !c_in_empty && !c_out_empty) {
+      // claccical vertex, can have bool in and out edges
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(c_in_ports == c_out_ports);  // bijection between in and out ports
+    } else if (!q_in_empty && !q_out_empty && !c_in_empty && !c_out_empty) {
+      // mixed vertex, can have bool in and out edges
+      CHECK(w_in_empty);
+      CHECK(w_out_empty);
+      CHECK(c_in_ports == c_out_ports);  // bijection between in and out ports
+      CHECK(q_in_ports == q_out_ports);  // bijection between in and out ports
+      // this case should only happen for measur circuits
```

Comment:
```suggestion
      // this case should only happen for measure ops
```

Reply:
done in [suggested changes I](https://github.com/CQCL/tket/pull/743/commits/e06543a3343027c8ff962d77fda395d3b22d9f96)  [e06543a](https://github.com/CQCL/tket/pull/743/commits/e06543a3343027c8ff962d77fda395d3b22d9f96)